lib/config-win32.h

```diff
@@ -609,6 +609,11 @@ Vista
 #  endif
 #endif
 
+/* Define if you have the ConnectEx WinSock 2 extension (Win 10 or newer) */
+#if defined(WINVER) && (WINVER >= 0x0A00)
+#define HAVE_CONNECTEX 1
```

Comment:
Shouldn't this be a runtime check? I mean, in Git for Windows, for example, we build cURL on one machine, but it is run on an estimated 3 million different machines.

Reply:
This check is specifically meant to verify you're building with a version of the Windows SDK that's new enough to support requesting the ConnectEx() function pointer from the OS. In fact, this specific check is a little too broad and is really superfluous when compared with the change to `lib/curl_setup.h` (which checks for a specific definition, `WSAID_CONNECTEX`).

There's a second check - at runtime (in `lib/easy.c`) - to ensure that the OS actually supports and returns this function pointer.

Reply:
Sounds like what you actually want is `_WIN32_WINNT_MAXVER` instead of `WINVER`.

Reply:
You should check if the operating system supports TCP Fast Open at runtime.

I looked at the PR code, actually in a project like Git for Windows (git relies on libcurl), even if it runs on Windows 10 1809, git can't open TCP Fast Open. The WIN_VER for these projects might be _WIN32_WINNT_WIN7

Reply:
> Sounds like what you actually want is `_WIN32_WINNT_MAXVER` instead of `WINVER`

Not really; I was assuming `WINVER` would equate to `_WIN32_WINNT`, but if anything, it should be the latter to exactly match the `#if` check in `MSWSock.h`:

```c
#if(_WIN32_WINNT >= 0x0501)
// ...
typedef
BOOL
(PASCAL FAR * LPFN_CONNECTEX) (
    _In_ SOCKET s,
    _In_reads_bytes_(namelen) const struct sockaddr FAR *name,
    _In_ int namelen,
    _In_reads_bytes_opt_(dwSendDataLength) PVOID lpSendBuffer,
    _In_ DWORD dwSendDataLength,
    _Out_ LPDWORD lpdwBytesSent,
    _Inout_ LPOVERLAPPED lpOverlapped
    );

#define WSAID_CONNECTEX \
    {0x25a207b9,0xddf3,0x4660,{0x8e,0xe9,0x76,0xe5,0x8c,0x74,0x06,0x3e}}
```

If you build with a new-enough Windows SDK but specifically target an earlier version of windows (by defining `WINVER`, `_WIN32_WINNT`, etc. with lower values than the MAXVER), you won't have those two definitions above.

Reply:
> You should check if the operating system supports TCP Fast Open at runtime.

Agreed - that's already done (see my first reply above).

> I looked at the PR code, actually in a project like Git for Windows (git relies on libcurl), even if it runs on Windows 10 1809, git can't open TCP Fast Open. The WIN_VER for these projects might be _WIN32_WINNT_WIN7

Doesn't it make sense it wouldn't work? TCP Fast Open wasn't supported in Windows 7, so if you're defining the constants to ensure the Windows headers only include features supported in Windows 7, you won't have the typedef/#define above that are only supported in later versions of Windows.

The only workaround would seem to be to duplicate the typedef/#define in the project's headers in a way that won't collide with the ones in the Windows headers. But that seems messy and in contradiction to the purpose of defining `_WINVER` / `_WIN32_WINNT` to values representing earlier versions of Windows.